File: app/utils/io.py
"""File I/O utilities."""

import json
from pathlib import Path
from typing import Any, Dict
from datetime import datetime
from app.config import RUNS_DIR


def create_run_directory(run_id: str) -> Path:
    """Create a directory for the current run."""
    run_dir = RUNS_DIR / run_id
    run_dir.mkdir(parents=True, exist_ok=True)
    return run_dir


def save_artifact(run_id: str, filename: str, content: Any, as_json: bool = False) -> Path:
    """Save an artifact to the run directory."""
    run_dir = create_run_directory(run_id)
    filepath = run_dir / filename
    
    if as_json:
        with open(filepath, 'w', encoding='utf-8') as f:
            if hasattr(content, 'model_dump'):
                json.dump(content.model_dump(), f, indent=2, default=str)
            else:
                json.dump(content, f, indent=2, default=str)
    else:
        with open(filepath, 'w', encoding='utf-8') as f:
            f.write(str(content))
    
    return filepath


def load_artifact(run_id: str, filename: str, as_json: bool = False) -> Any:
    """Load an artifact from the run directory."""
    filepath = RUNS_DIR / run_id / filename
    
    if not filepath.exists():
        raise FileNotFoundError(f"Artifact not found: {filepath}")
    
    if as_json:
        with open(filepath, 'r', encoding='utf-8') as f:
            return json.load(f)
    else:
        with open(filepath, 'r', encoding='utf-8') as f:
            return f.read()


def save_final_package(run_id: str, package: Any) -> Dict[str, Path]:
    """Save all final package artifacts."""
    saved_files = {}
    
    # Save JSON package
    saved_files['package'] = save_artifact(
        run_id, 'final_package.json', package, as_json=True
    )
    
    # Save brief
    saved_files['brief'] = save_artifact(
        run_id, 'brief.json', package.brief, as_json=True
    )
    
    # Save research
    saved_files['research'] = save_artifact(
        run_id, 'research.json', package.research_report, as_json=True
    )
    
    # Save blog draft
    blog_md = format_blog_markdown(package.blog_draft)
    saved_files['blog'] = save_artifact(run_id, 'blog.md', blog_md)
    
    # Save SEO blog
    saved_files['seo_blog'] = save_artifact(
        run_id, 'seo_blog.md', package.seo_optimized_blog.optimized_content
    )
    
    # Save social posts
    social_md = format_social_markdown(package.social_assets)
    saved_files['social'] = save_artifact(run_id, 'social.md', social_md)
    
    # Save email
    email_md = format_email_markdown(package.email_assets)
    saved_files['email'] = save_artifact(run_id, 'email.md', email_md)
    
    return saved_files


def format_blog_markdown(blog: Any) -> str:
    """Format blog draft as markdown."""
    lines = [
        f"# {blog.title}\n",
        "## Introduction\n",
        f"{blog.introduction}\n",
    ]
    
    for section in blog.body_sections:
        heading = section.heading if hasattr(section, 'heading') else section['heading']
        content = section.content if hasattr(section, 'content') else section['content']
        lines.append(f"## {heading}\n")
        lines.append(f"{content}\n")
    
    lines.append("## Conclusion\n")
    lines.append(f"{blog.conclusion}\n")
    lines.append(f"\n**{blog.call_to_action}**\n")
    lines.append(f"\n---\nWord count: {blog.word_count}")
    
    return "\n".join(lines)


def format_social_markdown(social: Any) -> str:
    """Format social assets as markdown."""
    lines = ["# Social Media Posts\n"]
    
    for post in social.posts:
        lines.append(f"## {post.platform}\n")
        lines.append(f"{post.content}\n")
        lines.append(f"**Hashtags:** {' '.join(post.hashtags)}\n")
        lines.append(f"*Character count: {post.character_count}*\n")
    
    if social.content_calendar_notes:
        lines.append(f"\n## Content Calendar Notes\n{social.content_calendar_notes}\n")
    
    return "\n".join(lines)


def format_email_markdown(email: Any) -> str:
    """Format email assets as markdown."""
    return f"""# Email Campaign

## Subject Line
{email.subject_line}

## Preview Text
{email.preview_text}

## Email Body
{email.email_body}

## Call to Action
{email.cta_text}
"""

File: src/lib/io.ts
import { saveArtifact as saveFile, ensureRunDir } from './logging';
import fs from 'fs/promises';
import path from 'path';

export async function saveFinalPackage(runId: string, pkg: any) {
  const saved: Record<string,string> = {};

  // Save final package JSON
  saved['package'] = await saveFile(runId, 'final_package.json', JSON.stringify(pkg, null, 2));

  // Save brief
  saved['brief'] = await saveFile(runId, 'brief.json', JSON.stringify(pkg.brief || {}, null, 2));

  // Save research
  saved['research'] = await saveFile(runId, 'research.json', JSON.stringify(pkg.research || {}, null, 2));

  // Save blog markdown
  const blogMd = formatBlogMarkdown(pkg.blog || pkg.blog_draft);
  saved['blog'] = await saveFile(runId, 'blog.md', blogMd);

  // Save seo optimized content
  const seoContent = (pkg.seo && pkg.seo.optimized_content) || (pkg.seo_optimized_blog && pkg.seo_optimized_blog.optimized_content) || '';
  saved['seo_blog'] = await saveFile(runId, 'seo_blog.md', seoContent || '');

  // Save social markdown
  const socialMd = formatSocialMarkdown(pkg.social || pkg.social_assets);
  saved['social'] = await saveFile(runId, 'social.md', socialMd);

  // Save email markdown
  const emailMd = formatEmailMarkdown(pkg.email || pkg.email_assets);
  saved['email'] = await saveFile(runId, 'email.md', emailMd);

  return saved;
}

export function formatBlogMarkdown(blog: any) {
  if (!blog) return '';
  const lines: string[] = [];
  lines.push(`# ${blog.title || ''}\n`);
  lines.push(`## Introduction\n`);
  lines.push(`${blog.introduction || ''}\n`);
  (blog.body_sections || []).forEach((section: any) => {
    const heading = section.heading || section.title || '';
    const content = section.content || section.body || '';
    lines.push(`## ${heading}\n`);
    lines.push(`${content}\n`);
  });
  lines.push('## Conclusion\n');
  lines.push(`${blog.conclusion || ''}\n`);
  if (blog.call_to_action) lines.push(`**${blog.call_to_action}**\n`);
  if (blog.word_count) lines.push(`\n---\nWord count: ${blog.word_count}`);
  return lines.join('\n');
}

export function formatSocialMarkdown(social: any) {
  if (!social || typeof social !== 'object') return '';
  const lines: string[] = [];
  lines.push('# Social Media Posts\n');
  
  const posts = social.posts || [];
  if (Array.isArray(posts) && posts.length > 0) {
    posts.forEach((post: any) => {
      lines.push(`## ${post.platform || 'Unknown Platform'}\n`);
      lines.push(`${post.content || ''}\n`);
      if (post.hashtags) {
        const tags = Array.isArray(post.hashtags) ? post.hashtags.join(' ') : String(post.hashtags);
        if (tags) lines.push(`**Hashtags:** ${tags}\n`);
      }
      if (post.character_count) lines.push(`*Character count: ${post.character_count}*\n`);
      lines.push('');
    });
  }

  if (social.content_calendar_notes) lines.push(`## Content Calendar Notes\n${social.content_calendar_notes}\n`);
  
  return lines.join('\n').trim();
}

export function formatEmailMarkdown(email: any) {
  if (!email || typeof email !== 'object') return '';
  const lines: string[] = [];
  lines.push('# Email Campaign\n');
  
  if (email.subject_line) lines.push(`## Subject Line\n${email.subject_line}\n`);
  if (email.preview_text) lines.push(`## Preview Text\n${email.preview_text}\n`);
  if (email.email_body) lines.push(`## Email Body\n${email.email_body}\n`);
  if (email.cta_text) lines.push(`## Call to Action\n${email.cta_text}\n`);
  
  return lines.join('\n').trim();
}
